language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -7,115 +7,65 @@ module Language.PlutusCore.TypeSynthesis ( typecheckProgram
                                          , normalizeType
                                          , runTypeCheckM
                                          , TypeCheckM
-                                         , BuiltinTable (..)
                                          , TypeError (..)
                                          , TypeCheckCfg (..)
                                          ) where
 
 import           Control.Monad.Except
 import           Control.Monad.Reader
 import           Control.Monad.State.Class
-import           Control.Monad.Trans.State      hiding (get, modify)
-import qualified Data.IntMap                    as IM
-import qualified Data.Map                       as M
+import           Control.Monad.Trans.State          hiding (get, modify)
+import qualified Data.IntMap                        as IM
 import           Language.PlutusCore.Clone
+import           Language.PlutusCore.Constant.Typed (typeOfBuiltinName)
 import           Language.PlutusCore.Error
-import           Language.PlutusCore.Lexer.Type hiding (name)
+import           Language.PlutusCore.Lexer.Type     hiding (name)
 import           Language.PlutusCore.Name
+import           Language.PlutusCore.Pretty         (prettyPlcDefString)
 import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Renamer        (annotateType)
 import           Language.PlutusCore.Type
 import           Lens.Micro
 import           PlutusPrelude
 
--- | A builtin table contains the kinds of builtin types and the types of
--- builtin names.
-data BuiltinTable = BuiltinTable (M.Map TypeBuiltin (Kind ())) (M.Map BuiltinName (NormalizedType TyNameWithKind ()))
+newtype TypeConfig = TypeConfig
+    { _typeConfigNormalizeTypes :: Bool  -- ^ Whether we normalize type annotations
```

Comment:
I believe we do *want* the typechecker to be parametric in a user-supplied builtin table, no? 

There's a ticket here: https://iohk.myjetbrains.com/youtrack/issue/CGP-222 which I believe still holds. We can use `typeOfBuiltin` if we'd like but then we should be parametric in the function for lookups. 

Also the field name should just be `_normalize` or `_typeConfigNormalize` if we want the prefix.

Reply:
@effectfully As @vmchale writes, we need to be able to pass a builtin table to the type checker that is used *in addition* the the builtins that are hardcoded (= in the spec). The reason that we want this to be a parameter is because we want to use the same type checker for PLC running on different blockchains (e.g., one with extended UTxO and one without). In these different contexts, the information that PLC code can get from the blockchain varies.

Reply:
I understand what you're saying, but I don't get what I should do. I mean, `M.Map BuiltinName (NormalizedType TyNameWithKind ())` can't be the right thing as `BuiltinName` is a non-extensive ADT, so you can't put there additional information (whether you're on a blockchain with extended UTxO or not).

> the information that PLC code can get from the blockchain varies.

This sounds to me like we should have additional machinery that allows to parse, type check and evaluate various extensions to the set of built-ins. But that's a huge task, why should I only change the type checker here in some ad hoc way?

Reply:
We can pass a function for lookups rather than a `Map`, no?

I believe the modifications to the parser would be relatively simple - and in any case we want to avoid regressions.

Reply:
> We can pass a function for lookups rather than a Map, no?

Lookups of what? I'm not asking for the type signature just to puzzle you, I really do not understand what I'm supposed to do.

> I believe the modifications to the parser would be relatively simple - and in any case we want to avoid regressions.

Let me repeat: the `M.Map BuiltinName (NormalizedType TyNameWithKind ())` thing makes zero sense when there is a method that allows to return a `NormalizedType TyNameWithKind ())` for each `BuiltinName` provided the set of `BuiltinName`s is fixed, which it is, because it's a non-extensible ADT. It's not a regression to remove something that makes zero sense.

Reply:
@mchakravarty, any clarification on this? This PR is several days old, but I still don't get what I'm supposed to change.

Reply:
So what I am proposing is that you take a function of type `BuiltinName -> Quote (Type TyName ())` (that is, the type of `typeOfBuiltinName`) as a parameter which is stored in the reader monad. Manuel can clarify if he has any better ideas.

Reply:
Well, without such a change it would be a regression. We can make `BuiltinName` extensible later on.

Reply:
That'd make our master in a very strange state where we pass unused environments around. It would make much more sense to implement the whole extensible built-in names story in a single PR instead of keeping a part here and adding a part later.

So I can do that, but it looks nasty to me. @mchakravarty, please command.

Reply:
@effectfully I agree that having `table <- defaultTable` as part of `runTypeCheckM` doesn't give us all the flexibility we need, because it effectively hardcodes the table into the type check. What we need instead is to parameterise `typecheckTerm` and `typecheckProgram` with a builtin table or with a lookup function. In other words, we want to include either `M.Map BuiltinName (NormalizedType TyNameWithKind ())` or a function of type `BuiltinName -> Quote (Type TyName ())` into `TypeCheckCfg`. My understanding of @vmchale's comments was that she was advocating the latter, which I think makes sense, as we can just pass `typeOfBuiltinName` in the case where we don't extend the table.

Does that clarify the issue?

BTW, three nitpicks:
* We have `typecheckTerm`, but `runTypeCheckM`. I think, we should use camelcase in a consistent manner.
* Also, we have `TypeCheckCfg` and `TypeConfig`. I think, abbreviations should also be used consistently.
* The three record type `TypeConfig`, `TypeCheckSt `, and `TypeCheckCfg ` use different conventions for prefixing field names. I think, the best thing is to just bite the bullet and use the *entire* type name as a prefix for the record field. They may get a bit longer, but at least are easy to remember. In cases where the type name is a very long (I mean, really very long), use an abbreviations formed from all the capital letters in the type name as the prefix.

@michaelpj Should we start adding things like the above to `CONTRIBUTING.md`?

Reply:
@mchakravarty I think that belongs to a separate Haskell style guide. Ideally one shared with the rest of IOHK!

Also, I think people aren't seeing Roman's point. `BuiltinName` is a closed ADT, *not* a type for which users can provide additional value, so taking a lookup function from it would only allow people to redefine the current set of builtins, which doesn't seem very useful. So having support for that here when we clearly haven't decided how we're going to represent extensible sets of builtins seems premature.

Reply:
> BTW, three nitpicks: ...

Yes, I specifically avoided doing any renaming in this PR. I agree with everything and 'll prepare another PR for this.

> Does that clarify the issue?

Michael said it better that I could:

> BuiltinName is a closed ADT, not a type for which users can provide additional value, so taking a lookup function from it would only allow people to redefine the current set of builtins, which doesn't seem very useful. So having support for that here when we clearly haven't decided how we're going to represent extensible sets of builtins seems premature.

Reply:
> I specifically avoided doing any renaming in this PR. I agree with everything and I'll prepare another PR for this.

Perfect!

Re dynamic builtins: I agree, the current set up is too inflexible and there is no point in having a partial solution, where it is unclear whether it can be extended to a full solution. Hence, I created a new YT item, which outlines the issue in detail: https://iohk.myjetbrains.com/youtrack/issue/CGP-374

We should look at this soon.